### PR TITLE
Opensource Notebook versioning based on Git

### PR DESCRIPTION
**Pull request by vidma**
This opensources support of notebook revision history. **Still Experimental**.
- `Filesystem` stays the default notebook storage, as it was
- `Git mode`  works as follows:
  * each edit (manual or autosave, if contents changed) becomes a git commit in a `local git dir`
  * if using a remote Git server (e.g. Github), it is cloned into the `local git dir` and each change commit is force pushed to the remote
- supported authentication methods:
  * http(s) - password or github token
  * SSH identity_file (password protection supported)
- it also allows to define other custom "storage providers"

see [git-notebook-provider/README](https://github.com/kensuio/spark-notebook/blob/d357b40e8b89682bf5d672c47cd5868bbd116140/modules/git-notebook-provider/README.md)

limitations to keep in mind: 
- each edit autosave currently creates a git commit. 
- config is slightly cumbersome. if git was configured incorrectly `spark-notebook` needs to be restarted, and it may be needed to clear the `local git dir` sometimes

git_revision_history

----

Changes:
- [x]  add abstract NotebookProviders interface
- [x] add FileSystemNotebookProvider
- [x] add Git notebook storage ( git-notebook-provider module)
  * see [git/README](https://github.com/kensuio/spark-notebook/blob/d357b40e8b89682bf5d672c47cd5868bbd116140/modules/git-notebook-provider/README.md) for details and configuration
  * see [conf/application-git-storage.conf](https://github.com/kensuio/spark-notebook/blob/d357b40e8b89682bf5d672c47cd5868bbd116140/conf/application-git-storage.conf) for example of how to configure it (environment variables are optional)
  * there is a [test-git-support.bash](https://github.com/kensuio/spark-notebook/blob/d357b40e8b89682bf5d672c47cd5868bbd116140/test-git-support.bash) which runs the tests on your git repo (config picked up from env variables)
- [x] add [run-dev.sh](https://github.com/kensuio/spark-notebook/blob/d357b40e8b89682bf5d672c47cd5868bbd116140/run-dev.sh) script which picks the right config
- [x] need to test if ssh auth with identity file (still) works
- [x] mention Git in the main README
- [x] hide revision-history when using a provider that do not support versions 231d702

cc @andypetrella @maasg

### CODE REVIEW

File: app/controllers/Application.scala [hunk 1]
```diff
@@ -400,12 +386,8 @@ object Application extends Controller {
   }
 
   def newDirectory(path: String, name:String) = {
```
Comment (maasg) [inline app/controllers/Application.scala]: maybe add return type for readability purposes?

File: app/controllers/Application.scala [hunk 2]
```diff
@@ -500,24 +482,26 @@ object Application extends Controller {
   }
 
   def listCheckpoints(snb: String) = Action { request =>
```
Comment (maasg) [inline app/controllers/Application.scala]: return type?

File: app/notebook/server/NotebookManager.scala [hunk 3]
```diff
@@ -85,82 +98,111 @@ class NotebookManager(val name: String, val notebookDir: File) {
...
+  }
+
   def getNotebook(path: String) = {
     Logger.info(s"getNotebook at path $path")
     for (notebook <- load(path)) yield {
       val data = FileUtils.readFileToString(notebookFile(path), "UTF-8")
-      val df = new SimpleDateFormat("dd-MM-yyyy HH:mm:ss z'('Z')'")
-      val last_mtime = df.format(new Date(notebookFile(path).lastModified()))
-      (last_mtime, notebook.name, data, path)
+      val lastModified = NotebookInfo.formatTimestamp(new Date(notebookFile(path).lastModified()))
```
Comment (maasg) [inline app/notebook/server/NotebookManager.scala]: maybe `formatTimestamp` could also take a `Long` as argument?

File: app/notebook/server/NotebookManager.scala [hunk 4]
```diff
@@ -85,82 +98,111 @@ class NotebookManager(val name: String, val notebookDir: File) {
...
+  }
 
 }
 
 class NotebookExistsException(message: String) extends IOException(message)
+class DirectoryCreationException(path: File) extends Exception("Could not create dir at: " + path.getCanonicalPath)
+
+case class NotebookInfo(lastModified: String, name: String, data: String, path: String)
+object NotebookInfo {
+  val dateTimeFormat = new SimpleDateFormat("dd-MM-yyyy HH:mm:ss z'('Z')'")
```
Comment (maasg) [inline app/notebook/server/NotebookManager.scala]: ohh. A shared evil `SimpleDateFormat`. This stuff is not threadsafe and will give us headaches. Consider alternatives (like jodatime or java.time) or at least instantiate it in the method below to have fresh :hankey: on each call.

File: build.sbt [hunk 5]
```diff
@@ -231,14 +227,25 @@ lazy val sparkNotebookCore = Project(id = "spark-notebook-core", base = file("mo
...
-  .aggregate(sparkNotebookCore, sbtDependencyManager, subprocess, observable, common, spark, kernel)
-  .dependsOn(sparkNotebookCore, subprocess, observable, common, spark, kernel)
+  .aggregate(sparkNotebookCore, gitNotebookProvider, sbtDependencyManager, subprocess, observable, common, spark, kernel)
+  .dependsOn(sparkNotebookCore, gitNotebookProvider, subprocess, observable, common, spark, kernel)
   .settings(sharedSettings: _*)
   .settings(
-    bashScriptExtraDefines <+= (version, scalaBinaryVersion, scalaVersion, sparkVersion, hadoopVersion, withHive) map { (v, sbv, sv, pv, hv, wh) =>
-      """export ADD_JARS="${ADD_JARS},${lib_dir}/$(ls ${lib_dir} | grep common.common | head)""""
+    bashScriptExtraDefines <+= (organization, version, scalaBinaryVersion, scalaVersion, sparkVersion, hadoopVersion, withHive) map { (org, v, sbv, sv, pv, hv, wh) =>
+      s"""export ADD_JARS="$${ADD_JARS},$${lib_dir}/$$(ls $${lib_dir} | ${org}.common | head)""""
```
Comment (maasg) [inline build.sbt]: I guess this expression: `ls $${lib_dir} | ${org}.common | head` will give some "command not found" error 
I think the `grep` is missing here when compared to the old version.

File: modules/core/src/main/scala/notebook/io/NotebookProviderFactory.scala [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+      .flatMap(identity)
+      .recoverWith { case ex: Exception =>
+        logError(s"Unable to instantiate provider from key [$providerConfiguratorClass]. Reason: [${ex.getMessage}]")
+        Future.failed(ex)
+      }
+
+    Await.result(eventualProvider, new FiniteDuration(initTimeoutMillis, MILLISECONDS))
+  }
+
+}
```
Comment (maasg) [inline modules/core/src/main/scala/notebook/io/NotebookProviderFactory.scala]: `<mark>`

Comment (vidma) [inline modules/core/src/main/scala/notebook/io/NotebookProviderFactory.scala]: ?

Comment (vidma) [inline modules/core/src/main/scala/notebook/io/NotebookProviderFactory.scala]: what do you mean here?

File: app/controllers/Application.scala (hunk 2 above)
Comment (vidma) [inline app/controllers/Application.scala]: other functions don't have it. so this should be a separate refactoring PR.

**Comment by vidma** [comment]
ready to 🚢

**Pull request merged**